File: src/diver2d.f
!                    ******************
                     SUBROUTINE DIVER2D
!                    ******************
     & (DUDX,DUDY,DVDX,DVDY,UO,VO,BOUND,UPBOUNDI,DOBOUNDI,LEBOUNDI,
     &  RIBOUNDI)

!
!***********************************************************************
! 2D-NAVIER STOKES SOLVER - FINITE DIFFERENCES
!***********************************************************************
!
!BRIEF    1) COMPUTE DIFFUSION COEFFICIENTS AND MATRIX.
!
!HISTORY  SERGIO CASTIBLANCO
!+        25/01/2021
!+        TRANSLATION FOR ORIGINAL MATLAB IMPLEMENTATION
!~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
!!
!! THIS FUNCTION COMPUTES THE TERMS DU/DX, DU/DY, DV/DX, DV/DY WITH 
!! UPWIND SECOND ORDER SCHEME (IF POSIBLE)
!!
!! NONLINEAL ADVECTION EQUATION
!!
!!   DU/DT + U DU/DX + V DU/DY = 0
!!
!!   DV/DT + U DV/DX + V DV/DY = 0
!!
!!%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
!! ENTRIES
!!   UO, VECTOR WITH VELOCITIES IN X DIRECTION
!!   UV, VECTOR WITH VELOCITIES IN Y DIRECTION
!!   DX, DIFFERENTIAL IN X
!!   DY, DIFFERENTIAL IN Y
!!   N, NUMBER OF NODES IN X DIRECTION
!!   M, NUMBER OF NODES IN Y DIRECTION
!!   BOUND, INDICES OF THE BOUNDARY NODES (B)
!!   UPBOUNDIN, INDICES WITH THE TOP INTERNAL BOUNDARY (U)
!!   DOBOUNDIN, INDICES WITH THE BOTTOM INTERNAL BOUNDARY (D)
!!   LEBOUNDIN, INDICES WITH THE LEFT INTERNAL BOUNDARY (L)
!!   RIBOUNDIN, INDICES WITH THE RIGHT INTERNAL BOUNDARY (R)
!!
!! GRID AND WHAT BOUND, UPBOUNDIN, DOBOUNDIN, LEBOUNDIN, AND RIBOUNDIN 
!! MEANS
!!
!! ENUMARTION USED IN THE GRID
!!
!!    25   26   27   28   29   30
!!
!!    19   20   21   22   23   24
!!
!!    13   14   15   16   17   18
!!
!!    7    8    9    10   11   12
!!
!!    1    2    3    4    5    6
!!
!! BOUND(B), UPBOUNDIN(U), DOBOUNDIN(D), LEBOUNDIN(L), RIBOUNDIN(R),
!! INTERNAL NODES(I)
!!
!!    B    B    B    B    B    B
!!
!!    B    UL   U    U    UR   B
!!
!!    B    L    I    I    R    B
!!
!!    B    DL   D    D    DR   B
!!
!!    B    B    B    B    B    B
!!
!! UL --> TOP-LEFT INTERNAL
!! DL --> BOTTOM-LEFT INTERNAL
!! DR --> BOTTOM-RIGHT INTERNAL
!! UR --> TOP-RIGHT INTERNAL
!!
!! THE RESULTS IN THE BOUNDARY ARE SET TO ZERO, BECAUSE THEY ARE NO 
!! NEEDED TO SOLVE NON-LINEAL ADVECTION EQUATION (DUE TO THE BOUNDARY 
!! CONDITIONS) IN OTHER WORDS, THE DIVERGENCE IS ONLY COMPUTED IN THE 
!! INTERNAL NODES.!
!!
!~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
!| DUDX      |<--| DU/DX VECTOR                                        |
!| DUDY      |<--| DU/DY VECTOR                                        |
!| DVDX      |<--| DV/DX VECTOR                                        |
!| DVDY      |<--| DV/DY VECTOR                                        |
!| UO        |-->| INITIAL/PREVIOUS VELOCITY IN X                      |
!| VO        |-->| INITIAL/PREVIOUS VELOCITY IN Y                      |
!| BOUND     |-->| BOUNDARY INDICES                                    |
!| UPBOUNDI  |-->| TOP INTERNAL BOUNDARY INDICES                       |
!| DOBOUNDI  |-->| BOTTOM INTERNAL BOUNDARY INDICES                    |
!| LEBOUNDI  |-->| LEFT INTERNAL BOUNDARY INDICES                      |
!| RIBOUNDI  |-->| RIGHT INTERNAL BOUNDARY INDICES                     |
!~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
!
      USE DECLARATIONS_NUMERICAL, ONLY: NX,NY,DX,DY
!
      IMPLICIT NONE
!
!~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
!
      DOUBLE PRECISION,DIMENSION(NX*NY),INTENT(IN) :: UO,VO
      DOUBLE PRECISION,DIMENSION(NX*NY),INTENT(OUT) :: DUDX,DUDY,
     &    DVDX,DVDY
      INTEGER, DIMENSION(2*NX+2*(NY-2)), INTENT(IN) :: BOUND
      INTEGER, DIMENSION(NX-4), INTENT(IN) :: UPBOUNDI, DOBOUNDI
      INTEGER, DIMENSION(NY-4), INTENT(IN) :: LEBOUNDI, RIBOUNDI
!
!  IN SUBROUTINE VARIABLES
!
      INTEGER :: I
!
!~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
!
!  DERIVATIVES IN THE INTERNAL NODES
!
      DO I=1,NX*NY
        IF((UO(I).GT.0).AND.(VO(I).GT.0)) THEN
          IF(ANY(I.EQ.BOUND)) THEN
            CONTINUE
          ELSEIF(ANY(I.EQ.DOBOUNDI).OR.(I.EQ.(2*NX-1)))THEN
            DUDX(I) = (1/(2*DX))*(3*UO(I) - 4*UO(I-1) + UO(I-2));
            DUDY(I) = (1/DY)*(UO(I) - UO(I-NX));
            DVDX(I) = (1/(2*DX))*(3*VO(I) - 4*VO(I-1) + VO(I-2));
            DVDY(I) = (1/DY)*(VO(I) - VO(I-NX));
          ELSEIF(ANY(I.EQ.LEBOUNDI).OR.(I.EQ.(NX*NY-2*NX+2))) THEN
            DUDX(I) = (1/DX)*(UO(I) - UO(I-1));
            DUDY(I) = (1/(2*DY))*(3*UO(I) - 4*UO(I-NX) + UO(I-2*NX));
            DVDX(I) = (1/DX)*(VO(I) - VO(I-1));
            DVDY(I) = (1/(2*DY))*(3*VO(I) - 4*VO(I-NX) + VO(I-2*NX));
          ELSEIF(I.EQ.(NX+2)) THEN
            DUDX(I) = (1/DX)*(UO(I) - UO(I-1));
            DUDY(I) = (1/DY)*(UO(I) - UO(I-NX));
            DVDX(I) = (1/DX)*(VO(I) - VO(I-1));
            DVDY(I) = (1/DY)*(VO(I) - VO(I-NX));
          ELSE
            DUDX(I) = (1/(2*DX))*(3*UO(I) - 4*UO(I-1) + UO(I-2));
            DUDY(I) = (1/(2*DY))*(3*UO(I) - 4*UO(I-NX) + UO(I-2*NX));
            DVDX(I) = (1/(2*DX))*(3*VO(I) - 4*VO(I-1) + VO(I-2));
            DVDY(I) = (1/(2*DY))*(3*VO(I) - 4*VO(I-NX) + VO(I-2*NX));
          ENDIF
        ELSEIF((UO(I).LE.0).AND.(VO(I).GT.0)) THEN
          IF(ANY(I.EQ.BOUND)) THEN
            CONTINUE
          ELSEIF(ANY(I.EQ.DOBOUNDI).OR.(I.EQ.(NX+2))) THEN
            DUDX(I) = (1/(2*DX))*(-3*UO(I) + 4*UO(I+1) - UO(I+2));
            DUDY(I) = (1/DY)*(UO(I) - UO(I-NX));
            DVDX(I) = (1/(2*DX))*(-3*VO(I) + 4*VO(I+1) - VO(I+2));
            DVDY(I) = (1/DY)*(VO(I) - VO(I-NX));
          ELSEIF(ANY(I.EQ.RIBOUNDI).OR.(I.EQ.(NX*NY-NX-1))) THEN
            DUDX(I) = (1/DX)*(-UO(I) + UO(I+1));
            DUDY(I) = (1/(2*DY))*(3*UO(I) - 4*UO(I-NX) + UO(I-2*NX));
            DVDX(I) = (1/DX)*(-VO(I) + VO(I+1));
            DVDY(I) = (1/(2*DY))*(3*VO(I) - 4*VO(I-NX) + VO(I-2*NX));
          ELSEIF(I.EQ.(2*NX-1)) THEN
            DUDX(I) = (1/DX)*(-UO(I) + UO(I+1));
            DUDY(I) = (1/DY)*(UO(I) - UO(I-NX));
            DVDX(I) = (1/DX)*(-VO(I) + VO(I+1));
            DVDY(I) = (1/DY)*(VO(I) - VO(I-NX));
          ELSE
            DUDX(I) = (1/(2*DX))*(-3*UO(I) + 4*UO(I+1) - UO(I+2));
            DUDY(I) = (1/(2*DY))*(3*UO(I) - 4*UO(I-NX) + UO(I-2*NX));
            DVDX(I) = (1/(2*DX))*(-3*VO(I) + 4*VO(I+1) - VO(I+2));
            DVDY(I) = (1/(2*DY))*(3*VO(I) - 4*VO(I-NX) + VO(I-2*NX));
          ENDIF
        ELSEIF((UO(I).GT.0).AND.(VO(I).LE.0)) THEN
          IF(ANY(I.EQ.BOUND)) THEN
            CONTINUE
          ELSEIF(ANY(I.EQ.UPBOUNDI).OR.(I.EQ.(NX*NY-NX-1))) THEN
            DUDX(I) = (1/(2*DX))*(3*UO(I) - 4*UO(I-1) + UO(I-2));
            DUDY(I) = (1/DY)*(-UO(I) + UO(I+NX));
            DVDX(I) = (1/(2*DX))*(3*VO(I) - 4*VO(I-1) + VO(I-2));
            DVDY(I) = (1/DY)*(-VO(I) + VO(I+NX));
          ELSEIF(ANY(I.EQ.LEBOUNDI).OR.(I.EQ.(NX+2))) THEN
            DUDX(I) = (1/DX)*(UO(I) - UO(I-1));
            DUDY(I) = (1/(2*DY))*(-3*UO(I) + 4*UO(I+NX) - UO(I+2*NX));
            DVDX(I) = (1/DX)*(VO(I) - VO(I-1));
            DVDY(I) = (1/(2*DY))*(-3*VO(I) + 4*VO(I+NX) - VO(I+2*NX));
          ELSEIF(I.EQ.(NX*NY-2*NX+2)) THEN
            DUDX(I) = (1/DX)*(UO(I) - UO(I-1));
            DUDY(I) = (1/DY)*(-UO(I) + UO(I+NX));
            DVDX(I) = (1/DX)*(VO(I) - VO(I-1));
            DVDY(I) = (1/DY)*(-VO(I) + VO(I+NX));
          ELSE
            DUDX(I) = (1/(2*DX))*(3*UO(I) - 4*UO(I-1) + UO(I-2));
            DUDY(I) = (1/(2*DY))*(-3*UO(I) + 4*UO(I+NX) - UO(I+2*NX));
            DVDX(I) = (1/(2*DX))*(3*VO(I) - 4*VO(I-1) + VO(I-2));
            DVDY(I) = (1/(2*DY))*(-3*VO(I) + 4*VO(I+NX) - VO(I+2*NX));
          ENDIF
        ELSE
          IF(ANY(I.EQ.BOUND)) THEN
            CONTINUE
          ELSEIF(ANY(I.EQ.UPBOUNDI).OR.(I.EQ.(NX*NY-2*NX+2))) THEN
            DUDX(I) = (1/(2*DX))*(-3*UO(I) + 4*UO(I+1) - UO(I+2));
            DUDY(I) = (1/DY)*(-UO(I) + UO(I+NX));
            DVDX(I) = (1/(2*DX))*(-3*VO(I) + 4*VO(I+1) - VO(I+2));
            DVDY(I) = (1/DY)*(-VO(I) + VO(I+NX));
          ELSEIF(ANY(I.EQ.RIBOUNDI).OR.(I.EQ.(2*NX-1))) THEN
            DUDX(I) = (1/DX)*(-UO(I) + UO(I+1));
            DUDY(I) = (1/(2*DY))*(-3*UO(I) + 4*UO(I+NX) - UO(I+2*NX));
            DVDX(I) = (1/DX)*(-VO(I) + VO(I+1));
            DVDY(I) = (1/(2*DY))*(-3*VO(I) + 4*VO(I+NX) - VO(I+2*NX));
          ELSEIF(I.EQ.(NX*NY-NX-1)) THEN
            DUDX(I) = (1/DX)*(-UO(I) + UO(I+1));
            DUDY(I) = (1/DY)*(-UO(I) + UO(I+NX));
            DVDX(I) = (1/DX)*(-VO(I) + VO(I+1));
            DVDY(I) = (1/DY)*(-VO(I) + VO(I+NX));
          ELSE
            DUDX(I) = (1/(2*DX))*(-3*UO(I) + 4*UO(I+1) - UO(I+2));
            DUDY(I) = (1/(2*DY))*(-3*UO(I) + 4*UO(I+NX) - UO(I+2*NX));
            DVDX(I) = (1/(2*DX))*(-3*VO(I) + 4*VO(I+1) - VO(I+2));
            DVDY(I) = (1/(2*DY))*(-3*VO(I) + 4*VO(I+NX) - VO(I+2*NX));
          ENDIF
        ENDIF
      ENDDO
!
!~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
!
      END SUBROUTINE DIVER2D

File: matlab/FTLE/diver2D.m
function [dudx,dudy,dvdx,dvdy] = diver2D(Uo,Vo,dx,dy,n,m,bound,...
        upboundin,doboundin,leboundin,riboundin)
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% This function computes the terms du/dx, du/dy, dv/dx, dv/dy with upwind
% second order scheme (if posible)
%
% Nonlineal advection equation
%
%   du/dt + u du/dx + v du/dy = 0
%
%   dv/dt + u dv/dx + v dv/dy = 0
%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Entries
%   Uo, vector with velocities in X direction
%   Uv, vector with velocities in Y direction
%   dx, differential in X
%   dy, differential in Y
%   n, number of nodes in X direction
%   m, number of nodes in Y direction
%   bound, indices of the boundary nodes (b)
%   upboundin, indices with the top internal boundary (u)
%   doboundin, indices with the bottom internal boundary (d)
%   leboundin, indices with the left internal boundary (l)
%   riboundin, indices with the right internal boundary (r)
%
% Grid and what bound, upboundin, doboundin, leboundin, and riboundin means
%
% Enumartion used in the grid
%
%    25   26   27   28   29   30
%
%    19   20   21   22   23   24
%
%    13   14   15   16   17   18
%
%    7    8    9    10   11   12
%
%    1    2    3    4    5    6
%
% bound(b), upboundin(u), doboundin(d), leboundin(l), riboundin(r),
% internal nodes(i)
%
%    b    b    b    b    b    b
%
%    b    ul   u    u    ur   b
%
%    b    l    i    i    r    b
%
%    b    dl   d    d    dr   b
%
%    b    b    b    b    b    b
%
% ul --> top-left internal
% dl --> bottom-left internal
% dr --> bottom-right internal
% ur --> top-right internal
%
% The results in the boundary are set to zero, because they are no needed
% to solve non-lineal advection equation (due to the boundary conditions)
% In other words, the divergence is only computed in the internal nodes.
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
   
    %Allocating memory for the result
    dudx = zeros(n*m,1);
    dudy = zeros(n*m,1);
    dvdx = zeros(n*m,1);
    dvdy = zeros(n*m,1);
    
    %
    %Derivatives in the internal nodes
    %
    for i=1:n*m
        if Uo(i) > 0 && Vo(i) > 0  %which means backward upwind in both directions
            %[doboundin, bottom-right internal node], second order in x, first order in y
            %[leboundin, top-left internal node], first order in x, second order in y
            %[bottom left internal node], first order in both directions
            %[riboundin,upboundin,top-right internal node, pure internal nodes], second order in x and y
            if ismember(i,bound)
                continue
            elseif (ismember(i,doboundin) || i==(2*n-1)) %doboundin, bottom-right internal node
                 dudx(i) = (1/(2*dx))*(3*Uo(i) - 4*Uo(i-1) + Uo(i-2));
                 dudy(i) = (1/dy)*(Uo(i) - Uo(i-n));
                 dvdx(i) = (1/(2*dx))*(3*Vo(i) - 4*Vo(i-1) + Vo(i-2));
                 dvdy(i) = (1/dy)*(Vo(i) - Vo(i-n));
            elseif (ismember(i,leboundin) || i==(n*m-2*n+2)) %leboundin, top-left internal node
                 dudx(i) = (1/dx)*(Uo(i) - Uo(i-1));
                 dudy(i) = (1/(2*dy))*(3*Uo(i) - 4*Uo(i-n) + Uo(i-2*n));
                 dvdx(i) = (1/dx)*(Vo(i) - Vo(i-1));
                 dvdy(i) = (1/(2*dy))*(3*Vo(i) - 4*Vo(i-n) + Vo(i-2*n));
            elseif i==(n+2) %bottom left internal node
                 dudx(i) = (1/dx)*(Uo(i) - Uo(i-1));
                 dudy(i) = (1/dy)*(Uo(i) - Uo(i-n));
                 dvdx(i) = (1/dx)*(Vo(i) - Vo(i-1));
                 dvdy(i) = (1/dy)*(Vo(i) - Vo(i-n));
            else %riboundin,upboundin,top-right internal node, pure internal nodes
                 dudx(i) = (1/(2*dx))*(3*Uo(i) - 4*Uo(i-1) + Uo(i-2));
                 dudy(i) = (1/(2*dy))*(3*Uo(i) - 4*Uo(i-n) + Uo(i-2*n));
                 dvdx(i) = (1/(2*dx))*(3*Vo(i) - 4*Vo(i-1) + Vo(i-2));
                 dvdy(i) = (1/(2*dy))*(3*Vo(i) - 4*Vo(i-n) + Vo(i-2*n));
            end
        elseif Uo(i) <= 0 && Vo(i) > 0  %which means forward upwind in x, and backward upwind in y
            %[doboundin, bottom-left internal node], second order in x, first order in y
            %[riboundin, top-right internal node], first order in x, second order in y
            %[bottom-right internal node], first order in both directions
            %[leboundin,upboundin,top-left internal node, pure internal nodes], second order in x and y
            if ismember(i,bound)
                continue
            elseif (ismember(i,doboundin) || i==(n+2)) %[doboundin, bottom-left internal node]
                dudx(i) = (1/(2*dx))*(-3*Uo(i) + 4*Uo(i+1) - Uo(i+2));
                dudy(i) = (1/dy)*(Uo(i) - Uo(i-n));
                dvdx(i) = (1/(2*dx))*(-3*Vo(i) + 4*Vo(i+1) - Vo(i+2));
                dvdy(i) = (1/dy)*(Vo(i) - Vo(i-n));
            elseif (ismember(i,riboundin) || i==(n*m-n-1)) %[riboundin, top-right internal node]
                dudx(i) = (1/dx)*(-Uo(i) + Uo(i+1));
                dudy(i) = (1/(2*dy))*(3*Uo(i) - 4*Uo(i-n) + Uo(i-2*n));
                dvdx(i) = (1/dx)*(-Vo(i) + Vo(i+1));
                dvdy(i) = (1/(2*dy))*(3*Vo(i) - 4*Vo(i-n) + Vo(i-2*n));
            elseif i==(2*n-1) %[bottom-right internal node]
                dudx(i) = (1/dx)*(-Uo(i) + Uo(i+1));
                dudy(i) = (1/dy)*(Uo(i) - Uo(i-n));
                dvdx(i) = (1/dx)*(-Vo(i) + Vo(i+1));
                dvdy(i) = (1/dy)*(Vo(i) - Vo(i-n));
            else %[leboundin,upboundin,top-left internal node, pure internal nodes]
                dudx(i) = (1/(2*dx))*(-3*Uo(i) + 4*Uo(i+1) - Uo(i+2));
                dudy(i) = (1/(2*dy))*(3*Uo(i) - 4*Uo(i-n) + Uo(i-2*n));
                dvdx(i) = (1/(2*dx))*(-3*Vo(i) + 4*Vo(i+1) - Vo(i+2));
                dvdy(i) = (1/(2*dy))*(3*Vo(i) - 4*Vo(i-n) + Vo(i-2*n));
            end
        elseif Uo(i) > 0 && Vo(i) <= 0  %which means backward upwind in x, and forward upwind in y
            %[upboundin, top-right internal node], second order in x, first order in y
            %[leboundin, bottom-left internal node], first order in x, second order in y
            %[top-left internal node], first order in both directions
            %[riboundin,doboundin,bottom-right internal node, pure internal nodes], second order in x and y
            if ismember(i,bound)
                continue
            elseif (ismember(i,upboundin) || i==(n*m-n-1)) %[upboundin, top-right internal node]
                dudx(i) = (1/(2*dx))*(3*Uo(i) - 4*Uo(i-1) + Uo(i-2));
                dudy(i) = (1/dy)*(-Uo(i) + Uo(i+n));
                dvdx(i) = (1/(2*dx))*(3*Vo(i) - 4*Vo(i-1) + Vo(i-2));
                dvdy(i) = (1/dy)*(-Vo(i) + Vo(i+n));
            elseif (ismember(i,leboundin) || i==(n+2)) %[leboundin, bottom-left internal node]
                dudx(i) = (1/dx)*(Uo(i) - Uo(i-1));
                dudy(i) = (1/(2*dy))*(-3*Uo(i) + 4*Uo(i+n) - Uo(i+2*n));
                dvdx(i) = (1/dx)*(Vo(i) - Vo(i-1));
                dvdy(i) = (1/(2*dy))*(-3*Vo(i) + 4*Vo(i+n) - Vo(i+2*n));
            elseif i==(n*m-2*n+2) %[top-left internal node]
                dudx(i) = (1/dx)*(Uo(i) - Uo(i-1));
                dudy(i) = (1/dy)*(-Uo(i) + Uo(i+n));
                dvdx(i) = (1/dx)*(Vo(i) - Vo(i-1));
                dvdy(i) = (1/dy)*(-Vo(i) + Vo(i+n));
            else %[riboundin,doboundin,bottom-right internal node, pure internal nodes]
                dudx(i) = (1/(2*dx))*(3*Uo(i) - 4*Uo(i-1) + Uo(i-2));
                dudy(i) = (1/(2*dy))*(-3*Uo(i) + 4*Uo(i+n) - Uo(i+2*n));
                dvdx(i) = (1/(2*dx))*(3*Vo(i) - 4*Vo(i-1) + Vo(i-2));
                dvdy(i) = (1/(2*dy))*(-3*Vo(i) + 4*Vo(i+n) - Vo(i+2*n));
            end
        else %which means forward upwind in both directions
            %[upboundin, top-left internal node], second order in x, first order in y
            %[riboundin, bottom-right internal node], first order in x, second order in y
            %[top-right internal node], first order in both directions
            %[leboundin,doboundin,bottom-left internal node, pure internal nodes], second order in x and y
            if ismember(i,bound)
                continue
            elseif (ismember(i,upboundin) || i==(n*m-2*n+2)) %[upboundin, top-left internal node]
                dudx(i) = (1/(2*dx))*(-3*Uo(i) + 4*Uo(i+1) - Uo(i+2));
                dudy(i) = (1/dy)*(-Uo(i) + Uo(i+n));
                dvdx(i) = (1/(2*dx))*(-3*Vo(i) + 4*Vo(i+1) - Vo(i+2));
                dvdy(i) = (1/dy)*(-Vo(i) + Vo(i+n));
            elseif (ismember(i,riboundin) || i==(2*n-1)) %[riboundin, bottom-right internal node]
                dudx(i) = (1/dx)*(-Uo(i) + Uo(i+1));
                dudy(i) = (1/(2*dy))*(-3*Uo(i) + 4*Uo(i+n) - Uo(i+2*n));
                dvdx(i) = (1/dx)*(-Vo(i) + Vo(i+1));
                dvdy(i) = (1/(2*dy))*(-3*Vo(i) + 4*Vo(i+n) - Vo(i+2*n));
            elseif i==(n*m-n-1) %[top-right internal node]
                dudx(i) = (1/dx)*(-Uo(i) + Uo(i+1));
                dudy(i) = (1/dy)*(-Uo(i) + Uo(i+n));
                dvdx(i) = (1/dx)*(-Vo(i) + Vo(i+1));
                dvdy(i) = (1/dy)*(-Vo(i) + Vo(i+n));
            else %[leboundin,doboundin,bottom-left internal node, pure internal nodes]
                dudx(i) = (1/(2*dx))*(-3*Uo(i) + 4*Uo(i+1) - Uo(i+2));
                dudy(i) = (1/(2*dy))*(-3*Uo(i) + 4*Uo(i+n) - Uo(i+2*n));
                dvdx(i) = (1/(2*dx))*(-3*Vo(i) + 4*Vo(i+1) - Vo(i+2));
                dvdy(i) = (1/(2*dy))*(-3*Vo(i) + 4*Vo(i+n) - Vo(i+2*n));
            end
        end
    end
            
end